e2e/internal/e2e/init/init_linux.go

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2019, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package init
+
+/*
+#define _GNU_SOURCE
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <sched.h>
+#include <sys/mount.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+
+#define SIZE    128
+
+// getProcInfo returns the parent PID, UID, and GID associated with the
+// supplied PID.
+static pid_t getProcInfo(pid_t pid, uid_t *uid, gid_t *gid) {
+	FILE *status;
+	char procPath[SIZE];
+	char *line = NULL;
+	size_t len = 0;
+	pid_t ppid = 1;
+
+	memset(procPath, 0, SIZE);
+	snprintf(procPath, SIZE-1, "/proc/%d/status", pid);
+
+	status = fopen(procPath, "r");
+	if ( status == NULL ) {
+		return 1;
+	}
+
+	while ( getline(&line, &len, status) != -1 ) {
+		if ( ppid == 1 ) {
+			sscanf(line, "PPid:\t%d", &ppid);
+		}
+		if ( *uid == 0 ) {
+			sscanf(line, "Uid:\t%d", uid);
+		}
+		if ( *gid == 0 ) {
+			sscanf(line, "Gid:\t%d", gid);
+		}
+	}
+
+	free(line);
+	fclose(status);
+
+	return ppid;
+}
+
+// getUnprivIDs searches recursively up the process parent chain to find a
+// process with a non-root UID, then returns the UID and GID of that process.
+static int getUnprivIDs(pid_t pid, uid_t *uid, gid_t *gid) {
+	if ( pid == 1 ) {
+		return -1;
+	}
+	pid_t ppid = getProcInfo(pid, uid, gid);
+	if ( *uid == 0 ) {
+		return getUnprivIDs(ppid, uid, gid);
+	}
+	return 0;
+}
+
+__attribute__((constructor)) static void init(void) {
+	uid_t uid = 0;
+	gid_t gid = 0;
+
+	if ( getUnprivIDs(getppid(), &uid, &gid) < 0 ) {
+		fprintf(stderr, "failed to retrieve user information\n");
+		exit(1);
+	}
+	if ( uid == 0 || gid == 0 ) {
+		fprintf(stderr, "failed to retrieve user information\n");
+		exit(1);
+	}
+
+	// use a mount namespace in order to bind a temporary filesystem
+	// on top of home directories and not screw them up by accident
+	// during tests execution
+	if ( unshare(CLONE_FS) < 0 ) {
+		fprintf(stderr, "failed to unshare filesystem: %s\n", strerror(errno));
+		exit(1);
+	}
+	if ( unshare(CLONE_NEWNS) < 0 ) {
```

Comment:
since this requires root (or CAP_SYS_ADMIN, same difference), maybe check for that before trying to do this in order to produce a better error message. The error message below is pretty generic and it will make things hard to debug for other people.

Reply:
will add a generic check at the beginning for root user